### PR TITLE
[SPARK-36721][SQL] Simplify boolean equalities if one side is literal

### PR DESCRIPTION
**Pull request by kazuyukitanimura**
### What changes were proposed in this pull request?
This PR proposes to improve simplifications of `EqualTo/EqualNullSafe` binary comparators when one side is a boolean literal.
For example: `EqualTo(predicate, TrueLiteral) => predicate`, `EqualNullSafe(predicate, TrueLiteral) => predicate if !predicate.nullable`
This PR helps pushing down the filter and reducing unnecessary IO.

### Why are the changes needed?
The following query does not push down the filter in the current implementation
```
SELECT * FROM t WHERE (a AND b) = true
```
although the following equivalent query pushes down the filter as expected.
```
SELECT * FROM t WHERE (a AND b)
```
That is because the first query creates `EqualTo(And(a, b), TrueLiteral)` that is simply not in the form that we can push down. However, we should be able to get it simplified to `And(a, b)`
It is fair for Spark SQL users to expect `(a AND b) = true` performs the same as `(a AND b)`.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added unit tests
```
build/sbt "testOnly *BooleanSimplificationSuite  -- -z SPARK-36721"
```

### CODE REVIEW

**Comment by kazuyukitanimura** [comment]
cc @cloud-fan @sunchao @viirya @dbtsai

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 1]
```diff
@@ -488,6 +489,16 @@ object SimplifyBinaryComparison
         // False with inequality
         case a GreaterThan b if canSimplifyComparison(a, b, notNullExpressions) => FalseLiteral
         case a LessThan b if canSimplifyComparison(a, b, notNullExpressions) => FalseLiteral
+
+        // One side is Literal
+        case (a: Predicate) EqualTo TrueLiteral => a
+        case TrueLiteral EqualTo (b: Predicate) => b
+        case (a: Predicate) EqualTo FalseLiteral => Not(a)
+        case FalseLiteral EqualTo (b: Predicate) => Not(b)
+        case (a: Predicate) EqualNullSafe TrueLiteral => And(a, IsNotNull(a))
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:498]: Hmm, doesn't it also mean the predicate `a` will be evaluated more than once?

Comment (kazuyukitanimura) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:498]: Thanks. I will change this to only for the case `!a.nullable`. I originally thought it is ok to evaluate twice in exchange of pushing down the filter. With the restriction of `!a.nullable`, the evaluation will be one time.

**Comment by kazuyukitanimura** [comment]
@viirya it would be great if you could take another look when you get a chance. Thank you!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala (hunk 1 above)
Comment (dbtsai) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: Let's say if an expression is not non-deterministic, evaluating it twice could return different results.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: Yea, deterministics is an issue. Performance maybe too, although we usually do subexpression elimination, but it possibly adds a bit extra cost. I think best practice should be not duplicate the evaluation.

**Comment by viirya** [comment]
retest this please

**Review by viirya — APPROVED** [review]
Looks okay. In case I may miss any point, I will leave this open for a while. If no more comment, I will merge this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 2]
```diff
@@ -488,6 +489,16 @@ object SimplifyBinaryComparison
         // False with inequality
         case a GreaterThan b if canSimplifyComparison(a, b, notNullExpressions) => FalseLiteral
         case a LessThan b if canSimplifyComparison(a, b, notNullExpressions) => FalseLiteral
+
+        // One side is Literal
+        case (a: Predicate) EqualTo TrueLiteral => a
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:494]: Does `a` have to be a `Predicate`? It seems like we can apply this optimization to any expression.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala (hunk 1 above)
Comment (kazuyukitanimura) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: Thank you all, updated

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala (hunk 2 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:494]: I guess the idea is for push down benefit. But for other expressions, it should be also okay.

Comment (kazuyukitanimura) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:494]: Thanks, yes the intention is for pushing down the filter. I will add a comment.
Actually, my original attempt, which I later changed to `Predicate` only, was `case a EqualTo TrueLiteral if a.dataType.isInstanceOf[BooleanType]`. I found that is too disruptive for some tests. I will keep it only for `Predicate` for now, and we can still get the push down benefit.

Comment (kazuyukitanimura) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:494]: updated

**Comment by cloud-fan** [comment]
> case a EqualTo TrueLiteral if a.dataType.isInstanceOf[BooleanType]

This is not necessary. The analyzer guarantees that the inputs of binary comparisons are of the same type. We can just write `case a EqualTo TrueLiteral`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/BinaryComparisonSimplificationSuite.scala [hunk 3]
```diff
@@ -194,4 +199,15 @@ class BinaryComparisonSimplificationSuite extends PlanTest with PredicateHelper
       Optimize.execute(testRelation.select(Coalesce(Seq('b, 'd)).as("out")).analyze),
       testRelation.select(Coalesce(Seq('b, 'd)).as("out")).analyze)
   }
+
+  test("SPARK-36721: Simplify boolean equalities if one side is literal") {
+    checkCondition(boolRelation, And('a, 'b) === TrueLiteral, And('a, 'b))
+    checkCondition(boolRelation, TrueLiteral === And('a, 'b), And('a, 'b))
+    checkCondition(boolRelation, And('a, 'b) === FalseLiteral, Or(Not('a), Not('b)))
+    checkCondition(boolRelation, FalseLiteral === And('a, 'b), Or(Not('a), Not('b)))
+    checkCondition(boolRelation, IsNull('a) <=> TrueLiteral, IsNull('a))
```
Comment (sunchao) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/BinaryComparisonSimplificationSuite.scala:208]: maybe ideally test the case when LHS is nullable but this is just a nit

**Review by sunchao — APPROVED** [review]
LGTM too!

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/BinaryComparisonSimplificationSuite.scala (hunk 3 above)
Comment (kazuyukitanimura) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/BinaryComparisonSimplificationSuite.scala:208]: Updated

**Comment by kazuyukitanimura** [comment]
> > case a EqualTo TrueLiteral if a.dataType.isInstanceOf[BooleanType]
> 
> This is not necessary. The analyzer guarantees that the inputs of binary comparisons are of the same type, and optimizer will never hit an unresolved plan. We can just write `case a EqualTo TrueLiteral`

Thanks, updated @cloud-fan

**Review by viirya — APPROVED** [review]
lgtm. See if @cloud-fan has more comments.

**Review by cloud-fan — APPROVED** [review]

**Comment by viirya** [comment]
Thanks all! Merging to master.

**Comment by dongjoon-hyun** [comment]
+1, LGTM, too.

**Pull request closed without merging**